coriolis/providers/replicator.py

```diff
@@ -283,6 +283,114 @@ def init_replicator(self):
             self._setup_replicator)(self._ssh)
         utils.retry_on_error()(
             self._init_replicator_client)(self._credentials)
+        LOG.debug(
+            "Disk status after Replicator initialization: %s",
+            self._cli.get_status(device=None, brief=True))
+
+    def get_current_disks_status(self):
+        """ Returns a list of the current status of the attached data disks.
+        The root disk of the worker VM is NOT returned.
+        The result is a list with elements of the following format:
+        [{
+            'device-path': '/dev/xvdf',
+            'device-name': 'xvdf',
+            'size': 10737418240,
+            'checksum-algorithm': 'sha256',
+            'chunk-size': 10485760,
+            'logical-sector-size': 512,
+            'physical-sector-size': 512,
+            'alignment-offset': 0,
+            'has-mounted-partitions': False,
+            'checksum-status': {
+                'status': 'running',
+                'total-chunks': 1024,
+                'completed-chunks': 0,
+                'percentage': 0},
+            'partitions': [{
+                'name': 'xvdf1',
+                'sectors': 20969439,
+                'uuid': '73e9659d-2fd9-46ca-a341-5a8637c416ee',
+                'fs': 'ext4',
+                'start-sector': 2048,
+                'end-sector': 20971486,
+                'alignment-offset': 0}]
+        }]
+        """
+        return self._cli.get_status()
+
+    def attach_new_disk(
+            self, disk_id, attachf, previous_disks_status=None):
+        """ Returns the volumes_info for the disk attached by running
+        `attachf`. This is achieved by comparing the disk statuses before and
+        after the execution of the attachment operation.
+
+        param disk_id: str(): the 'disk_id' of the info from self._volumes_info
+            for the disk which shall be attached by `attachf()`.
+        param attachf: fun(): argument-less function which attaches the disk.
+            The function should perform any waits required to reasonably expect
+            the worker OS to have noticed the new disk, or reboot the worker VM
+            and re-run `init_replicator()` if deemed necessary.
+        param previous_disks_status: dict(): previous status of the disks as
+            returned by `get_current_disks_status()`.
+
+        return: dict(): returns the volumes_info associated to the new disk.
+        """
+        # check if volume with given ID is present in self._volumes_info:
+        matching_vols = [
+            vol for vol in self._volumes_info
+            if vol['disk_id'] == disk_id]
+        if not matching_vols:
+            raise exception.CoriolisException(
+                "No information regarding volume with ID '%s'. "
+                "Cannot track its attachment." % disk_id)
+        elif len(matching_vols) > 2:
```

Comment:
Shouldn't this be ```!= 1```?

Also, this can be a standalone if, instead of an ```elif```

Reply:
Good catch, fixing now